Code/Framework/AzCore/Tests/Debug/PerformanceCollectorTests.cpp

```diff
@@ -0,0 +1,215 @@
+        // In real life these parameters are mirrored by CVARs.
+        const AZ::u32 numberOfCaptureBatches = 3;
+        const AZ::Debug::PerformanceCollector::DataLogType dataLogType = AZ::Debug::PerformanceCollector::DataLogType::LogStatistics;
+        const AZ::u32 waitTimePerCaptureBatch = 0; //0 seconds
+        const AZ::u32 frameCountPerCaptureBatch = 10;
+
+        AZ::u32 pendingBatchCount = numberOfCaptureBatches;
+        auto onCompleteCallback = [&](AZ::u32 batchCount) {
+            pendingBatchCount = batchCount;
+        };
+        AZStd::vector<AZStd::string_view> paramList = { PerfParam1, PerfParam2, PerfParam3 };
+        AZ::Debug::PerformanceCollector performanceCollector(LogCategory, paramList, onCompleteCallback);
+        performanceCollector.UpdateDataLogType(dataLogType);
+        performanceCollector.UpdateFrameCountPerCaptureBatch(frameCountPerCaptureBatch);
+        performanceCollector.UpdateWaitTimeBeforeEachBatch(AZStd::chrono::seconds(waitTimePerCaptureBatch));
+        performanceCollector.UpdateNumberOfCaptureBatches(numberOfCaptureBatches);
+
+        // We will loop more than is required to simulate OnSystemTick, but the amount of collected data should
+        // be limited to the specified control parameters.
+        const AZ::u32 EXCESS_FRAME_LOOP = 5;
+        for (AZ::u32 frameLoop = 0; frameLoop < ((numberOfCaptureBatches*frameCountPerCaptureBatch) + EXCESS_FRAME_LOOP); frameLoop++)
+        {
+            performanceCollector.FrameTick();
+            if (!performanceCollector.IsWaitingBeforeCapture())
+            {
+                performanceCollector.RecordPeriodicEvent(PerfParam1);
+            }
+            {
+                AZ::Debug::ScopeDuration param1Scope(&performanceCollector, PerfParam2);
+            }
+            {
+                AZ::Debug::ScopeDuration param1Scope(&performanceCollector, PerfParam3);
+            }
+        }
+
+        auto stringWithOutputData = performanceCollector.GetOutputDataBuffer();
+        EXPECT_FALSE(stringWithOutputData.empty());
+        EXPECT_EQ(pendingBatchCount, 0);
+
+        rapidjson::Document jsonDoc;
+        jsonDoc.Parse(stringWithOutputData.c_str());
+        ASSERT_TRUE(!jsonDoc.HasParseError());
+
+        // The root is an array of dictionaries.
+        ASSERT_TRUE(jsonDoc.IsArray());
+        // We have 3 paramaters and 3 capture batches. So we expect 9 arrays.
+        auto expectedArraySize = paramList.size() * numberOfCaptureBatches;
+        ASSERT_TRUE(jsonDoc.Size() == expectedArraySize);
+        for (rapidjson::SizeType i = 0; i < expectedArraySize; i += aznumeric_cast<rapidjson::SizeType>(paramList.size()))
+        {
+            AZStd::unordered_set<AZStd::string> validatedParams;
+            AZStd::string paramName = jsonDoc[i]["name"].GetString();
+            EXPECT_NE(AZStd::find(paramList.begin(), paramList.end(), paramName), paramList.end());
```

Comment:
FYI This can use `AZStd::ranges::contain` to return a boolean as to whether the paramName is not in the list
```suggestion
            EXPECT_FALSE(AZStd::ranges::contains(paramList, paramName));
```

Reply:
Love it, thanks. I knew there was a better way.

Reply:
I tried it, but didn't work. Looked at other code using ranges::contains and they use regular AZStd::vector.
When I do `auto paramList = AZStd::vector<AZStd::string>({ PerfParam1, PerfParam2, PerfParam3 });`   
it works, but when I do:  
`auto paramList = AZStd::to_array<AZStd::string_view>({ PerfParam1, PerfParam2, PerfParam3 });` I get this error:
```
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2672: 'operator __surrogate_func': no matching overloaded function found
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2672:             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2672:             ^
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2893: Failed to specialize function template 'enable_if<std::conjunction_v<std::integral_constant<bool,AZStd::ranges::input_range<T>>,std::integral_constant<bool,AZStd::indirect_binary_predicate<AZStd::ranges::equal_to,AZStd::projected<unknown-type,Proj,void>,const T*,void>>>,bool>::type AZStd::ranges::Internal::contains_fn::operator ()(R &&,const T &,Proj) const'
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2893:             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2893:             ^
1>D:\GIT\o3de\Code\Framework\AzCore\AzCore/std/ranges/ranges_algorithm.h(2028): message : see declaration of 'AZStd::ranges::Internal::contains_fn::operator ()'
1>D:\GIT\o3de\Code\Framework\AzCore\AzCore/std/ranges/ranges_algorithm.h(2028): message :             constexpr auto operator()(R&& r, const T& value, Proj proj = {}) const
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message : With the following template arguments:
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             ^
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message : 'R=AZStd::array<AZStd::basic_string_view<char,AZStd::char_traits<char>>,3> &'
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             ^
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message : 'T=AZStd::string'
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             ^
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message : 'Proj=AZStd::identity'
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             ^
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2512: 'testing::AssertionResult': no appropriate default constructor available
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2512:             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2512:             ^
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message : No constructor could take the source type, or constructor overload resolution was ambiguous
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): message :             ^
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2737: 'gtest_ar_': const object must be initialized
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2737:             EXPECT_TRUE(AZStd::ranges::contains(paramList, paramName));
1>D:\GIT\o3de\Code\Framework\AzCore\Tests\Debug\PerformanceCollectorTests.cpp(90,13): error C2737:             ^
```